src/plugins/filetracer/filetracer.cpp

```diff
@@ -103,940 +103,29 @@
+#include "linux.h"
+#include "win.h"
 
-static event_response_t create_file_cb(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+filetracer::filetracer(drakvuf_t drakvuf, output_format_t output)
 {
-    /*
-    __kernel_entry NTSTATUS NtCreateFile(
-      OUT PHANDLE           FileHandle,
-      IN ACCESS_MASK        DesiredAccess,
-      IN POBJECT_ATTRIBUTES ObjectAttributes,
-      OUT PIO_STATUS_BLOCK  IoStatusBlock,
-      IN PLARGE_INTEGER     AllocationSize,
-      IN ULONG              FileAttributes,
-      IN ULONG              ShareAccess,
-      IN ULONG              CreateDisposition,
-      IN ULONG              CreateOptions,
-      IN PVOID              EaBuffer,
-      IN ULONG              EaLength
-    );
-    */
-    filetracer* f = (filetracer*)info->trap->data;
-    struct wrapper* w = new wrapper;
-    if (!w) return 0;
-    w->f = f;
-    w->rsp = info->regs->rsp;
-
-    w->pid = info->attached_proc_data.pid;
-    if (!drakvuf_get_current_thread_id(drakvuf, info, &w->tid) && !w->tid)
-        PRINT_DEBUG("filetracer: Failed to get TID");
-
-    w->handle = drakvuf_get_function_argument(drakvuf, info, 1);
-    if ( !w->handle )
-    {
-        printf("filetracer: Null pointer pHandle in CreateFile\n");
-        delete w;
-        return VMI_EVENT_RESPONSE_NONE;
-    }
-    w->obj_attr = drakvuf_get_function_argument(drakvuf, info, 3);
-    auto attrs_value = drakvuf_get_function_argument(drakvuf, info, 6);
-    w->attrs = parse_flags(attrs_value, file_flags_and_attrs, f->format);
-    auto share_value = drakvuf_get_function_argument(drakvuf, info, 7);
-    w->share = parse_flags(share_value, share_mode, f->format);
-    auto disp_value = drakvuf_get_function_argument(drakvuf, info, 8);
-    w->disp = parse_flags(disp_value, disposition, f->format);
-    auto opts_value = drakvuf_get_function_argument(drakvuf, info, 9);
-    w->opts = parse_flags(opts_value, create_options, f->format);
-    auto access_value = drakvuf_get_function_argument(drakvuf, info, 2);
-    w->access = opts_value & FILE_DIRECTORY_FILE
-                ? parse_flags(access_value, directory_ar, f->format)
-                : parse_flags(access_value, file_ar, f->format);
-
-    addr_t rsp = 0;
-    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
-    vmi_read_addr_va(vmi, info->regs->rsp, 0, &rsp);
-    drakvuf_release_vmi(drakvuf);
-
-    drakvuf_trap_t* trap = (drakvuf_trap_t*)g_malloc0(sizeof(drakvuf_trap_t));
-    trap->breakpoint.lookup_type = LOOKUP_PID;
-    trap->breakpoint.pid = 4;
-    trap->breakpoint.addr_type = ADDR_VA;
-    trap->breakpoint.addr = rsp;
-    trap->type = BREAKPOINT;
-    trap->name = info->trap->name;
-    trap->data = w;
-    trap->cb = create_file_ret_cb;
-
-    if ( !drakvuf_add_trap(drakvuf, trap) )
+    this->os = drakvuf_get_os_type(drakvuf);
+    if (this->os == VMI_OS_WINDOWS)
     {
-        printf("Failed to trap return at 0x%lx\n", rsp);
-        delete w;
+        new win_filetracer(drakvuf, output);
```

Comment:
Valgrind reports memory leak, this never gets `delete`-d when shutting down the plugin